### PR TITLE
Function transformers

### PR DESCRIPTION
**Pull request by alvaro-castillo**

### CODE REVIEW

File: skfda/preprocessing/feature_construction/_function_transformers.py [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+
+        Then we decide how many intervals we want to consider (in our case 2)
+        and call the function with the dataset.
+        >>> import numpy as np
+        >>> from skfda.preprocessing.feature_construction import (
+        ...     LocalAveragesTransformer,
+        ... )
+        >>> local_averages = LocalAveragesTransformer(2)
+        >>> np.around(local_averages.fit_transform(X), decimals=2)
+        array([[[ 116.94],
```
Comment (vnmabus) [inline skfda/preprocessing/feature_construction/_function_transformers.py:40]: This should be a 2D matrix to be able to use it in further scikit-learn estimators.

File: skfda/preprocessing/feature_construction/_function_transformers.py [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+    >>> from scipy.special import jv
+    >>> from skfda.representation import FDataGrid
+    >>> import numpy as np
+    >>> x_grid = np.linspace(0, 14, 14)
+    >>> fd_grid = FDataGrid(
+    ...     data_matrix=[jv([0], x_grid)],
+    ...     grid_points=x_grid,
+    ... )
+    >>> fd_grid.data_matrix
+    array([[[ 1.        ],
```
Comment (vnmabus) [inline skfda/preprocessing/feature_construction/_function_transformers.py:161]: This should also need to be a 2D matrix.

**Review by vnmabus — CHANGES_REQUESTED** [review]

File: skfda/preprocessing/feature_construction/_function_transformers.py (hunk 2 above)
Comment (alvaro-castillo) [reply skfda/preprocessing/feature_construction/_function_transformers.py:161]: The output of OccupationMeasureTransformer is indeed a 2D array. What you are pointing is the data_matrix of a Bessel Function on a FDataGrid format.

Comment (vnmabus) [reply skfda/preprocessing/feature_construction/_function_transformers.py:157]: Sure thing! Sorry for the mistake.

File: skfda/exploratory/stats/_functional_transformers.py [hunk 3]
```diff
@@ -127,7 +67,7 @@ def local_averages(
         data.integrate(interval=((intervals[i], intervals[i + 1]))) / step
         for i in range(n_intervals)
     ]
-    return np.asarray(integrated_data)
+    return np.asarray(integrated_data)[:, :, 0]
```
Comment (vnmabus) [inline skfda/exploratory/stats/_functional_transformers.py:70]: I am actually fine with this function returning a ND-array. It is the transformer, the one which has to return a 2D array. And I think that has to be done using `reshape` and not getting only the 0-th element, as that won't work for more general functions.

File: skfda/exploratory/stats/_functional_transformers.py [hunk 4]
```diff
@@ -34,85 +34,25 @@ def local_averages(
         calculate the local averages.
         n_intervals: number of intervals we want to consider.
     Returns:
-        ndarray of shape (n_intervals, n_samples, n_dimensions)
-        with the transformed data for FDataBasis and (n_intervals, n_samples)
-        for FDataGrid.
+        ndarray of shape (n_intervals, n_samples) with the
```
Comment (vnmabus) [inline skfda/exploratory/stats/_functional_transformers.py:37]: I think that the first dimension should always be the number of samples, for consistency and compatibility with scikit-learn tools.

File: skfda/preprocessing/feature_construction/_function_transformers.py [hunk 5]
```diff
@@ -37,12 +37,8 @@ class LocalAveragesTransformer(BaseEstimator, TransformerMixin):
         ... )
         >>> local_averages = LocalAveragesTransformer(2)
         >>> np.around(local_averages.fit_transform(X), decimals=2)
-        array([[[ 116.94],
-                [ 111.86],
-                [ 107.29]],
-               [[ 177.26],
-                [ 157.62],
-                [ 154.97]]])
+        array([[ 116.94,  111.86,  107.29],
```
Comment (vnmabus) [inline skfda/preprocessing/feature_construction/_function_transformers.py:40]: The number of rows should always be the number of samples.

**Review by vnmabus — CHANGES_REQUESTED** [review]

File: skfda/preprocessing/feature_construction/_function_transformers.py (hunk 5 above)
Comment (alvaro-castillo) [reply skfda/preprocessing/feature_construction/_function_transformers.py:40]: `occupation_measure` then should also be changed to comply with shape(n_samples, n_intervals)

File: skfda/exploratory/stats/_functional_transformers.py [hunk 6]
```diff
@@ -127,7 +71,7 @@ def local_averages(
         data.integrate(interval=((intervals[i], intervals[i + 1]))) / step
         for i in range(n_intervals)
     ]
-    return np.asarray(integrated_data)
+    return np.transpose(integrated_data, (1, 0, 2))
 
 
 def _calculate_curves_occupation_(
```
Comment (vnmabus) [inline skfda/exploratory/stats/_functional_transformers.py:77]: Why the last underscore?

File: skfda/exploratory/stats/_functional_transformers.py [hunk 7]
```diff
@@ -254,7 +199,7 @@ def occupation_measure(
         function_y_coordinates,
         function_x_coordinates,
         intervals,
-    )
+    ).T
```
Comment (vnmabus) [inline skfda/exploratory/stats/_functional_transformers.py:202]: Couldn't the computation be done already using the right dimensions?

**Review by vnmabus — CHANGES_REQUESTED** [review]

**Pull request merged**